src/gui/properties/piecesbar.cpp

```diff
@@ -0,0 +1,305 @@
+        QColor highlightColor {this->palette().color(QPalette::Active, QPalette::Highlight)};
+        highlightColor.setAlphaF(0.35);
+        QRect targetHighlightRect {m_highlitedRegion.adjusted(borderWidth, borderWidth, borderWidth, height() - 2 * borderWidth)};
+        painter.fillRect(targetHighlightRect, highlightColor);
+    }
+
+    QPainterPath border;
+    border.addRect(0, 0, width(), height());
+    painter.setPen(m_borderColor);
+    painter.drawPath(border);
+}
+
+void PiecesBar::requestImageUpdate()
+{
+    if (updateImage(m_image))
+        update();
+}
+
+QRgb PiecesBar::mixTwoColors(QRgb rgb1, QRgb rgb2, float ratio)
+{
+    int r1 = qRed(rgb1);
+    int g1 = qGreen(rgb1);
+    int b1 = qBlue(rgb1);
+
+    int r2 = qRed(rgb2);
+    int g2 = qGreen(rgb2);
+    int b2 = qBlue(rgb2);
+
+    float ratio_n = 1.0 - ratio;
+    int r = (r1 * ratio_n) + (r2 * ratio);
+    int g = (g1 * ratio_n) + (g2 * ratio);
+    int b = (b1 * ratio_n) + (b2 * ratio);
+
+    return qRgb(r, g, b);
+}
+
+namespace {
+#define TYPESET_FILELIST_AS_TABLE
+
+    class DetailedTooltipRenderer
+    {
+    public:
+        DetailedTooltipRenderer(QTextStream& stream, const QString &header)
+            : m_stream(stream)
+        {
+            m_stream << header
+#ifdef TYPESET_FILELIST_AS_TABLE
+                << R"(<table style="width:100%; padding: 3px; vertical-align: middle;">)";
+#else
+                << "<ul>";
+#endif
+        }
+
+        ~DetailedTooltipRenderer()
+        {
+#ifdef TYPESET_FILELIST_AS_TABLE
+            m_stream << "</table>";
+#else
+            m_stream << "</ul>";
+#endif
+        }
+
+        void operator()(const QString &size, const QString &path)
+        {
+#ifdef TYPESET_FILELIST_AS_TABLE
+            m_stream << "<tr><td>" << size << "</td><td>" << path << "</td></tr>";
+#else
+            m_stream << "<li>[" << size << "] " << path << "</li>";
+#endif
+        }
+
+    private:
+        QTextStream &m_stream;
+    };
+}
+
+void PiecesBar::showToolTip(const QHelpEvent *e)
+{
+    if (!m_torrent)
+        return;
+
+    QString toolTipText;
+    QTextStream stream(&toolTipText, QIODevice::WriteOnly);
+    bool showDetailedInformation = QApplication::keyboardModifiers().testFlag(Qt::ShiftModifier);
+    if (!showDetailedInformation) {
+        if (m_local) {
+            stream << tr("White: Missing pieces") << '\n'
+                   << tr("Green: Partial pieces") << '\n'
+                   << tr("Blue: Completed pieces") << '\n';
+        }
+        else {
+            stream << tr("White: Unavailable pieces") << '\n'
+                   << tr("Blue: Available pieces") << '\n';
+        }
+        stream << '\n' << tr("Hold Shift key for detailed information");
+    }
+    else {
+        stream << "<html><body>";
+        const int imagePos = e->pos().x() - borderWidth;
+        if (( imagePos >=0) && ( imagePos < m_image.width()) ) {
+            PieceIndexToImagePos transform {m_torrent, m_image};
```

Comment:
Look, I just asked do you ban this particular feature or not. 

> And how it can help you here and in all similar code?

It helps when used systematically without thinking/decision. In the same way as you switch on a blinker before any turn and before even looking into a mirror, which helps you only when you make a mistake and overlook somebody in  the mirror. 

> When I asked you to discuss this rule, you arrogantly refused.

I don't care about coding style of a particular project I'm contributing too as far as the project provides a tool (or a tool config) to ensure/enforce that style. Whatever the style is, the main difficulties in understanding the another person code is not the number of spaces or position of braces, but, provided the code is well-written, different way of thinking of its authors. And this is unavoidable in OSS projects. 

As for the banning particular language features, I don't understand this entirely. If one thinks that a language feature is not needed, then, perhaps, there are languages which suit better?

I'm sorry that my refusal can be interpreted as a haughtiness, but in fact it was quite the opposite: I fell that I can do nothing on this subject. Starting from the fact that I'm in ultra-right tabs party, so will not like the qBt code style without changing it completely.